### PR TITLE
Fix implicit conversion from String to Datetime

### PR DESCRIPTION
**Pull request by riverszhang89**
To compare 2 operands of different types, one of the operands is first promoted to the type of the other operand. After the comparison, the changes to the flags on the promoted operand are undone, demoting the operand to its original type. However, there're a few issues in this logic, when it comes to comparing a String with a Datetime.

### Invalid datetime strings causing arbitrary results

Consider the following table and query

```
Table t

| col (datetime)|
|---------------|
|   2021-01-01  |
|   2021-01-01  |

SELECT * FROM t WHERE col < "ABC"
```

The literal "ABC" isn't a valid datetime. Therefore when converting it to Datetime, `sqlite3VdbeMemDatetimefy()` fails and returns an error. Unfortunately `sqlite3MemCompare()` ignores the error, calls `dttz_cmp()`, and ends up comparing an uninitialized `dttz_t` field in the Mem object of "ABC" with the Datetime column. The results are hence undefined.

The patch fixes it by checking for any conversion error and passing it back to the VDBE, so that such a query will yield an error instead of returning arbitrary results.

### Valid datetime strings working by coincidence

Consider the following table and query

```
Table t

| col (datetime)|
|---------------|
|   2021-01-01  |
|   2021-01-01  |

SELECT * FROM t WHERE col < '2022-01-01'
```

When comparing the String operand with the 1st Datetime value in the table, the operand is promoted to Datetime. `sqlite3VdbeMemDatetimefy()`, however, frees `Mem.z` and sets it to null. After the comparison, the operand's flags are restored to `MEM_Str`. At this point, we have a `MEM_Str` operand, with a null string value, and a `dttz_t` field of January 1st, 2022.

Now the operand is compared to the 2nd value. `sqlite3VdbeMemDatetimefy()` is going to fail, as the operand is not a legitimate string any more. `sqlite3MemCompare()`, as we mentioned, conveniently ignores the conversion error, compares the previously converted `dttz_t` field to the 2nd row value, and returns a correct result.

The logic here is obviously flawed. The patch improves it by keeping an operand's flags if it's converted from a String to a non-String type, and has lost its string value during the conversion. So in the case above, the String literal will be converted to a Datetime exactly once, and stays a Datetime for the rest comparisons.

### Datetime strings without timezone resulting in undefined row orders in order-by

The function `str_to_dttz()` expects a valid timezone. However, in an order-by query, MEM_Str objects deserialized from the VDBESorter do not have their `tz` set, causing `str_to_dttz()` to return an error. Again, because `compareDateTimeInterval()` does not check for any type error, it may cause the VDBESorter to return an undefined row order.

To fix it, we simply use the `tzname` from the `clnt`, if `tz` isn't present.

### CODE REVIEW

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ sqlite/src/vdbeInt.h @@
+ If the order of members in struct Mem is changed, columnNullValue (defined in sqlite/vdbeapi.c) needs to be reordered accordingly.
@@ sqlite/src/vdbeaux.c @@
+ If a new fixed-length serial type is introduced, sqlite3IsFixedLengthSerialType (defined in sqlite/vdbeInt.h) needs to be changed accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>3/401 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>simple_timepart_reptimeout_generated<br>sc_timepart_logicalsc_generated<br>simple_timepart</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/cd70c2eb/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/cd70c2eb/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2747'>Resubmit</a></p>

**Review by adizaimi — DISMISSED** [review]
The changes look correct to me!

File: sqlite/src/vdbe.c [hunk 1]
```diff
@@ -2522,6 +2522,11 @@ case OP_Ge: {             /* same as TK_GE, jump, in1, in3 */
 #endif /* defined(SQLITE_BUILDING_FOR_COMDB2) */
     assert( pOp->p4type==P4_COLLSEQ || pOp->p4.pColl==0 );
     res = sqlite3MemCompare(pIn3, pIn1, pOp->p4.pColl);
+    if( p->rc!=SQLITE_OK ){
+      rc = p->rc;
+      p->rc = 0;
+      goto abort_due_to_error;
+    }
```
Comment (nirbhayc) [inline sqlite/src/vdbe.c:2529]: Place this in SQLITE_BUILDING_FOR_COMDB2 block.

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ sqlite/src/vdbeInt.h @@
+ If the order of members in struct Mem is changed, columnNullValue (defined in sqlite/vdbeapi.c) needs to be reordered accordingly.
@@ sqlite/src/vdbeaux.c @@
+ If a new fixed-length serial type is introduced, sqlite3IsFixedLengthSerialType (defined in sqlite/vdbeInt.h) needs to be changed accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>4/401 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>simple_timepart_reptimeout_generated<br>sc_timepart<br>simple_timepart<br>logfill</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/d9c9cb64/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/d9c9cb64/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2747'>Resubmit</a></p>

**Review by mponomar — APPROVED** [review]

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ sqlite/src/vdbeInt.h @@
+ If the order of members in struct Mem is changed, columnNullValue (defined in sqlite/vdbeapi.c) needs to be reordered accordingly.
@@ sqlite/src/vdbeaux.c @@
+ If a new fixed-length serial type is introduced, sqlite3IsFixedLengthSerialType (defined in sqlite/vdbeInt.h) needs to be changed accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/c8e02db6/smoketest.log'>Error &#x26A0; </a>.<br>Failed to compile.</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/c8e02db6/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/c8e02db6/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2747'>Resubmit</a></p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ sqlite/src/vdbeInt.h @@
+ If the order of members in struct Mem is changed, columnNullValue (defined in sqlite/vdbeapi.c) needs to be reordered accordingly.
@@ sqlite/src/vdbeaux.c @@
+ If a new fixed-length serial type is introduced, sqlite3IsFixedLengthSerialType (defined in sqlite/vdbeInt.h) needs to be changed accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>6/401 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>simple_timepart_reptimeout_generated<br>sc_timepart_logicalsc_generated<br>sc_timepart<br>simple_timepart<br>sp_queueodh_generated<br>op_makerecord</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/9a9fea94/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/9a9fea94/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2747'>Resubmit</a></p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ sqlite/src/vdbeInt.h @@
+ If the order of members in struct Mem is changed, columnNullValue (defined in sqlite/vdbeapi.c) needs to be reordered accordingly.
@@ sqlite/src/vdbeaux.c @@
+ If a new fixed-length serial type is introduced, sqlite3IsFixedLengthSerialType (defined in sqlite/vdbeInt.h) needs to be changed accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>4/401 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>simple_timepart_reptimeout_generated<br>sc_timepart<br>simple_timepart<br>auth_twofiles_generated</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/9a9fea94/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/9a9fea94/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2747'>Resubmit</a></p>

**Review by roborivers — COMMENTED** [review]
Lint check:

```diff


@@ sqlite/src/vdbeInt.h @@
+ If the order of members in struct Mem is changed, columnNullValue (defined in sqlite/vdbeapi.c) needs to be reordered accordingly.
@@ sqlite/src/vdbeaux.c @@
+ If a new fixed-length serial type is introduced, sqlite3IsFixedLengthSerialType (defined in sqlite/vdbeInt.h) needs to be changed accordingly.
```

**Review by roborivers — CHANGES_REQUESTED** [review]
<p><a href='https://bloomberg.github.io/comdb2/contrib.html#code-style'>Coding style</a> check: Success &#10003;.<br>Smoke testing: Success &#10003;.<br>Cbuild submission: Success &#10003;.<br>Regression testing: <b>2/401 tests failed</b> &#x26A0;.</p><p>The first 10 failing tests are: <br>simple_timepart_reptimeout_generated<br>simple_timepart</p><p><a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/9a9fea94/detail.txt'>Detail</a> | <a href='http://comdb2.s3-website-us-east-1.amazonaws.com/tests/9a9fea94/logs.tar.gz'>Log</a> | <a href='http://ec2-34-202-216-70.compute-1.amazonaws.com/restricted/index.php?id=2747'>Resubmit</a></p>

File: sqlite/src/vdbeaux.c [hunk 2]
```diff
@@ -4848,7 +4875,51 @@ int sqlite3MemCompare(const Mem *pMem1, const Mem *pMem2, const CollSeq *pColl){
...
+      int rc = compareDateTimeInterval(f1, f2, combined_flags, pMem1, pMem2, &err);
+      if( err==0 )
+        return rc;
+      /* Type error. If one value is a string, convert the other one to
+         a string and then compare. */
+      if( f1&MEM_Str ){
+        sqlite3VdbeMemInit(&copy, pMem1->db, MEM_Null);
+        sqlite3VdbeMemCopy(&copy, pMem2);
+        sqlite3VdbeMemStringify(&copy, pColl ? pColl->enc : pMem1->enc, 0);
+
```
Comment (mistachkin) [inline sqlite/src/vdbeaux.c:4909]: What if this call to sqlite3VdbeMemStringify fails?

Comment (riverszhang89) [reply sqlite/src/vdbeaux.c:4912]: I added a check for the return code. Thanks for pointing it out!

**Review by mistachkin — APPROVED** [review]
Looks good to me.

**Comment by riverszhang89** [comment]
@mistachkin Thank you for the review!

**Pull request merged**